### ISSUE TITLE
Update mataa_impedance_fit_speaker.m line 138

### ISSUE THREAD

**Issue by estearg** [mbrennwa/mataa#8]
Hello Mathias,
I would follow the standard procedure of forking, modifying and making a pull request, but I believe all is needed is simply to change line 138 in mataa_impedance_fit_speaker.m, writing mataa_impedance_speaker_model_LR2 instead of mataa_impedance_speaker_model.
Best regards,
Esteban

**Comment by mbrennwa** [comment]
Thanks. You're right. I also changed mataa_impedance_fit_speaker.m to mataa_impedance_fit_speaker_LR2.m to avoid confusion with the WRIGHT impedance model. I haven't tested the revisions, though. Please let me know if there are issues.

**Comment by estearg** [comment]
Thank you very much Matt, there are no issues.
Anyway, I've found some time ago that the procedure gives Qms = 0 with valid input data. I then resort to R. H. Small, "Direct-Radiator Loudspeaker System Analysis", p. 276 (Measurement of Driver Parameters) and p. 282 (to apply equation 53 as a check). Does d'Appolito consider something else (than the classic method) for the low-frequency fit?
Always a pleasure to hear you.

**Comment by mbrennwa** [comment]
I currently have no access to these references. Can you somehow provide the equations and their descriptions? Also, and maybe more important, can you provide an example of data and code that yields Qms = 0?

**Comment by estearg** [comment]
Hello Matthias,

You can find both Thiele's and Small's papers at http://www.readresearch.co.uk/articles.php and at http://diyaudioprojects.com/Technical/ (I found them when reading https://www.andyc.diy-audio-engineering.org/vented_box_references.html).
The procedure using traditional instruments -which of course at some time I replaced with the computer's sound card and some programs I wrote-, is explained at https://sound-au.com/tsp.htm.

I'm attaching a file that yields Qms = 0. It has f, mag and phase variables returned by
[mag, phase, f] = mataa_measure_impedance(10, 1000, 9.9);
They look pretty good when plotted with
mataa_plot_impedance(mag, phase, f);
They are measurements of an ordinary speaker, a model whose specifications I also attach (it's in Spanish but the symbols are universal.)
I have two units and I could measure the other one using MATAA without problems, and with good agreement with the manual method. (I also verified the agreement between MATAA and the manual method with other speakers.)
When I run
[Rdc,f0,Qe,Qm,L1,L2,R2] = mataa_impedance_fit_speaker_LR2 (f,mag,phase)
I get Rdc = 6.5116, f0 = 78.612, Qe = 2.9399, Qm = 0 for the low-frequency part.
I had measured with the classic method Rdc = 6.3, f0 = 74.54, Qe = 0.748, Qm = 4.145, which are similar results to those of the other unit.

Best regards,

Esteban

[MN-6L.pdf](https://github.com/mbrennwa/mataa/files/3448080/MN-6L.pdf)
[MN-6L-2-Z-2017-10-16.txt](https://github.com/mbrennwa/mataa/files/3448036/MN-6L-2-Z-2017-10-16.txt)

**Comment by mbrennwa** [comment]
The issue is related to how the initial values for the fit(s) are determined. I found a way to make this work with your data example. However, I can't commit to GitHub at the moment, because I keep getting an error saying that my password is not valid (although it is). Let's see if this goes away after a while...

**Comment by estearg** [comment]
Hello Matthias,

I tried the new code and it works fine with that set of values.

The same file 'MN-6L-2-Z-2017-10-16.txt' contains another set of values, with the variables suffixed with the letter 'A', that correspond to the impedance measurement of the same speaker with 8.8 g added mass (to compute Vas).
The code applied to that set gives an f0 (f0A I call it) greater than without added mass (f0A = 84.596 Hz, f0 = 78.833 Hz), which does not agree with the measured impedance magnitude and phase (nor with what I measured manually). Plotting 'magA' and 'phaseA' alongside 'mag' and 'phase' it seems clear that the resonance has lowered its frequency, as expected.

I would like to help you, but I don't understand very well the reasoning behind the algorithm (and I don't have access to d'Appolito's book, in case it's related to it.)

Thank you very much for taking the time to see the issue.

Best regards,

Esteban

**Comment by mbrennwa** [comment]
I have to admit that I haven't used mataa_impedance_fit_speaker for a long time, mainly because it does not work as well as one might hope. I mostly just "fit" the parameters by hand.

mataa_impedance_fit_speaker works by first trying to figure out reasonable guesstimates of the parameters, which are then used as starting values for the fit. The code then first fits the low-frequency part (the main peak) of the impedance curve, to get a better guesstimate of the low-frequency parameters. Then it continues to the full fit of the full curve. The whole procedure only works out well if the values of the first guesstimates are sufficiently good; and they often are not. Maybe it would be better if the function would be changed such that the user specifies "good" starting values rather than attempting an automated guess. What do you think?

**Comment by estearg** [comment]
Hello Matthias,

I have to admit the same: I did not report the issue when it first
appeared because I continued with the classic method, so I apologize
for that.

Unfortunately I don't know much about curve fitting in GNU Octave/MATLAB.
Tomorrow I'll test leasqr with the low-frequency part of the impedance
and the classic model (which, if I'm not mistaken, is common to LR-2
and Wright models at low frequencies), if you have not already done
it.

If all fitting methods fail, there could be a separate function
applying equations to calculate the low-frequency parameters.

Best regards,

Esteban


2019-08-04 7:24 GMT-03:00, mbrennwa <notifications@github.com>:
